File: java/reflect/src/fan/reflect/Field.java
package fan.reflect;

import fan.std.Map;
import fan.sys.*;
import fanx.fcode.*;
import fanx.main.*;

/**
 * Field is a slot which "stores" a value.
 */
public class Field extends Slot {

	//////////////////////////////////////////////////////////////////////////
	// Factories
	//////////////////////////////////////////////////////////////////////////

	public static Func makeSetFunc(final Map map) {
		return new Func() {
			public Object call(Object obj) {
				List keys = map.keys();
				for (int i = 0; i < keys.size(); ++i) {
					Field field = (Field) keys.get(i);
					Object val = map.get(field);
					field.set(obj, val, obj != inCtor);
				}
				return null;
			}

			@Override
			public boolean isImmutable() {
				return true;
			}
		};
	}

	//////////////////////////////////////////////////////////////////////////
	// Java Constructor
	//////////////////////////////////////////////////////////////////////////

	public Field(Type parent, String name, int flags, List facets, int lineNum, Type type) {
		super(parent, name, flags, facets, lineNum);
		this.type = type;
	}
	
	public static Field fromFCode(FField f, Type parent) {
		//TODO
		List facets = List.make(Sys.findType("sys::Facet"), 1);
		FType ftype = parent.ftype();
		FTypeRef tref = ftype.pod.typeRef(f.type);
		Type type = Sys.findType(tref.signature);
		Field field = new Field(parent, f.name, f.flags, facets, 0, type);
		
		try {
			Class<?> clz = parent.getJavaClass();
			field.reflect = clz.getDeclaredField(f.name);
			field.reflect.setAccessible(true);
		} catch (NoSuchFieldException e) {
			e.printStackTrace();
		} catch (SecurityException e) {
			e.printStackTrace();
		}
		
		return field;
	}

	//////////////////////////////////////////////////////////////////////////
	// Signature
	//////////////////////////////////////////////////////////////////////////

	private static Type typeof = Sys.findType("reflect::Field");

	public Type typeof() {
		return typeof;
	}

	public Type type() {
		return type;
	}

	public String signature() {
		return type.signature() + " " + name;
	}

	public Object trap(String name, List args) {
		// private undocumented access
		if (name.equals("getter"))
			return getter;
		if (name.equals("setter"))
			return setter;
		return super.trap(name, args);
	}

	//////////////////////////////////////////////////////////////////////////
	// Reflection
	//////////////////////////////////////////////////////////////////////////

	public Object get() {
		return get(null);
	}

	public Object get(Object instance) {
		// parent.finish();

		if (getter != null) {
			return getter.invoke(instance, Method.noArgs);
		}

		try {
			// if JavaType handle slot resolution
			// if (parent.isJava()) return JavaType.get(this, instance);
			return reflect.get(instance);
		} catch (Exception e) {
			if (reflect == null)
				throw Err.make("Field not mapped to java.lang.reflect correctly " + qname());

			throw Err.make(e);
		}
	}

	public void set(Object instance, Object value) {
		set(instance, value, true);
	}

	public void set(Object instance, Object value, boolean checkConst) {
		// parent.finish();
		// check const
		if ((flags & FConst.Const) != 0) {
			if (checkConst)
				throw ReadonlyErr.make("Cannot set const field " + qname());
			else if (value != null && !isImmutable(value))
				throw ReadonlyErr.make("Cannot set const field " + qname() + " with mutable value");
		}

		// check static
		if ((flags & FConst.Static) != 0)
			throw ReadonlyErr.make("Cannot set static field " + qname());

		// // check generic type (the Java runtime will check non-generics)
		// if (type.isGenericInstance() && value != null)
		// {
		// if (!typeof(value).is(type.toNonNullable()))
		// throw ArgErr.make("Wrong type for field " + qname() + ": " + type + "
		// != " + typeof(value));
		// }

		// use the setter by default, however if we have a storage field and
		// the setter was auto-generated then falldown to set the actual field
		// to avoid private setter implicit overrides
		if ((setter != null && !setter.isSynthetic()) || reflect == null) {
			setter.invoke(instance, new Object[] { value });
			return;
		}

		try {
			// if JavaType handle slot resolution
			// if (parent.isJava()) { JavaType.set(this, instance, value);
			// return; }
			reflect.set(instance, value);
		} catch (IllegalArgumentException e) {
			throw ArgErr.make(e);
		} catch (Exception e) {
			if (reflect == null)
				throw Err.make("Field not mapped to java.lang.reflect correctly");

			throw Err.make(e);
		}
	}

	//////////////////////////////////////////////////////////////////////////
	// Fields
	//////////////////////////////////////////////////////////////////////////

	Type type;
	Method getter;
	Method setter;
	java.lang.reflect.Field reflect;
	Method overload; // if overloaded by method in JavaType

}

File: src/std/js/Field.js
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   24 Mar 09  Andy Frank  Creation
//   20 May 09  Andy Frank  Refactor to new OO model
//

/**
 * Field.
 */
fan.std.Field = fan.sys.Obj.$extend(fan.sys.Slot);

//////////////////////////////////////////////////////////////////////////
// Factories
//////////////////////////////////////////////////////////////////////////

fan.std.Field.makeSetFunc = function(map)
{
  return fan.sys.Func.make(
    fan.sys.List.make(fan.sys.Param.$type),
    fan.sys.Void.$type,
    function(obj)
    {
      var keys = map.keys();
      for (var i=0; i<keys.size(); i++)
      {
        var field = keys.get(i);
        var val = map.get(field);
        field.set(obj, val, false); //, obj != inCtor);
      }
    });
}

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

fan.std.Field.prototype.$ctor = function(parent, name, flags, type, facets)
{
  this.m_parent = parent;
  this.m_name   = name;
  this.m_qname  = parent.qname() + "." + name;
  this.m_flags  = flags;
  this.m_type   = type;
  this.m_$name  = this.$$name(name);
  this.m_$qname = this.m_parent.m_$qname + '.m_' + this.m_$name;
  this.m_getter = null;
  this.m_setter = null;
  this.m_facets = new fan.sys.Facets(facets);
}

//////////////////////////////////////////////////////////////////////////
// Obj
//////////////////////////////////////////////////////////////////////////

fan.std.Field.prototype.trap = function(name, args)
{
  // private undocumented access
  if (name == "getter") return this.m_getter;
  if (name == "setter") return this.m_setter;
  return fan.sys.Obj.prototype.trap.call(this, name, args);
}

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

fan.std.Field.prototype.type = function() { return this.m_type; }

fan.std.Field.prototype.get = function(instance)
{
  if (this.isStatic())
  {
    return eval(this.m_$qname);
  }
  else
  {
    var target = instance;
    if ((this.m_flags & fan.sys.FConst.Native) != 0)
      target = instance.peer;
    var getter = target[this.m_$name];
    if (getter != null)
      return getter.call(target);
    else
      return target["m_"+this.m_$name]
  }
}

fan.std.Field.prototype.set = function(instance, value, checkConst)
{
  if (checkConst === undefined) checkConst = true;

  // check const
  if ((this.m_flags & fan.sys.FConst.Const) != 0)
  {
    if (checkConst)
      throw fan.sys.ReadonlyErr.make("Cannot set const field " + this.m_qname);
    else if (value != null && !fan.sys.ObjUtil.isImmutable(value))
      throw fan.sys.ReadonlyErr.make("Cannot set const field " + this.m_qname + " with mutable value");
  }

  // check static
  if ((this.m_flags & fan.sys.FConst.Static) != 0) // && !parent.isJava())
    throw fan.sys.ReadonlyErr.make("Cannot set static field " + this.m_qname);

  // check type
  if (value != null && !fan.sys.ObjUtil.$typeof(value).is(this.m_type.toNonNullable()))
    throw fan.sys.ArgErr.make("Wrong type for field " + this.m_qname + ": " + this.m_type + " != " + fan.sys.ObjUtil.$typeof(value));

  // TODO
  //if (setter != null)
  //{
  //  setter.invoke(instance, new Object[] { value });
  //  return;
  //}

  if ((this.m_flags & fan.sys.FConst.Native) != 0)
  {
    var peer = instance.peer;
    var setter = peer[this.m_$name + "$"];
    setter.call(peer, instance, value);
  }
  else
  {
    var setter = instance[this.m_$name + "$"];
    if (setter != null)
      setter.call(instance, value);
    else
      instance["m_"+this.m_$name] = value;
  }
}

fan.std.Field.prototype.$typeof = function() { return fan.std.Field.$type; }

